File: Phase-02/Todo-Full-Stack/frontend/src/types/reminder.ts
/**
 * Reminder types for todo reminders.
 *
 * Task Reference: T005 - Create frontend types for reminders
 * Feature: 006-recurring-reminders
 */

/**
 * Status of a reminder.
 */
export type ReminderStatus = "pending" | "sent" | "snoozed" | "cancelled";

/**
 * Preset reminder offset values.
 */
export type ReminderOffsetPreset =
  | "at_time"
  | "5_min"
  | "15_min"
  | "30_min"
  | "1_hour"
  | "2_hours"
  | "1_day"
  | "2_days"
  | "1_week";

/**
 * Request to create a reminder.
 */
export interface ReminderCreate {
  /** Todo ID to attach reminder to */
  todo_id: string;
  /** Occurrence ID for recurring todos (optional) */
  occurrence_id?: string;
  /** Absolute time to fire the reminder (UTC ISO string) */
  fire_at?: string;
  /** Minutes before due time (negative number, mutually exclusive with fire_at) */
  offset_minutes?: number;
}

/**
 * Response for a reminder.
 */
export interface ReminderResponse {
  id: string;
  todo_id: string;
  occurrence_id: string | null;
  fire_at: string;
  offset_minutes: number;
  status: ReminderStatus;
  sent_at: string | null;
  snoozed_until: string | null;
  created_at: string;
}

/**
 * Request to snooze a reminder.
 */
export interface SnoozeRequest {
  /** Minutes from now to snooze */
  snooze_minutes?: number;
  /** Preset snooze duration */
  preset?: "5_minutes" | "15_minutes" | "1_hour" | "custom";
}

/**
 * Map of preset values to offset minutes.
 */
export const REMINDER_OFFSET_VALUES: Record<ReminderOffsetPreset, number> = {
  "at_time": 0,
  "5_min": -5,
  "15_min": -15,
  "30_min": -30,
  "1_hour": -60,
  "2_hours": -120,
  "1_day": -1440,
  "2_days": -2880,
  "1_week": -10080,
};

/**
 * Labels for reminder offset presets.
 */
export const REMINDER_OFFSET_LABELS: Record<ReminderOffsetPreset, string> = {
  "at_time": "At due time",
  "5_min": "5 minutes before",
  "15_min": "15 minutes before",
  "30_min": "30 minutes before",
  "1_hour": "1 hour before",
  "2_hours": "2 hours before",
  "1_day": "1 day before",
  "2_days": "2 days before",
  "1_week": "1 week before",
};

/**
 * Snooze duration options.
 */
export const SNOOZE_OPTIONS = [
  { value: 5, label: "5 minutes" },
  { value: 15, label: "15 minutes" },
  { value: 60, label: "1 hour" },
] as const;

/**
 * Format offset minutes to human-readable string.
 */
export function formatReminderOffset(offsetMinutes: number): string {
  const absMinutes = Math.abs(offsetMinutes);

  if (absMinutes < 60) {
    return `${absMinutes} minute${absMinutes !== 1 ? "s" : ""} before`;
  }

  const hours = Math.floor(absMinutes / 60);
  if (absMinutes < 1440) {
    return `${hours} hour${hours !== 1 ? "s" : ""} before`;
  }

  const days = Math.floor(absMinutes / 1440);
  if (absMinutes < 10080) {
    return `${days} day${days !== 1 ? "s" : ""} before`;
  }

  const weeks = Math.floor(absMinutes / 10080);
  return `${weeks} week${weeks !== 1 ? "s" : ""} before`;
}

/**
 * Get preset from offset minutes, if exact match exists.
 */
export function getPresetFromOffset(offsetMinutes: number): ReminderOffsetPreset | null {
  for (const [preset, value] of Object.entries(REMINDER_OFFSET_VALUES)) {
    if (value === offsetMinutes) {
      return preset as ReminderOffsetPreset;
    }
  }
  return null;
}

File: Phase-02/Todo-Full-Stack/backend/app/schemas/reminder.py
"""
Reminder request/response schemas.

Task Reference: T032 - Create ReminderCreate, ReminderResponse schemas
Feature: 006-recurring-reminders

Pydantic schemas for reminder management.
"""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, model_validator


class ReminderCreate(BaseModel):
    """
    Request schema for creating a reminder.

    Task Reference: T032
    """

    todo_id: str = Field(
        ...,
        description="Todo to attach the reminder to",
        examples=["todo_abc123"],
    )
    occurrence_id: Optional[str] = Field(
        default=None,
        description="Specific occurrence ID (for recurring todos)",
        examples=["occ_xyz789"],
    )
    fire_at: Optional[datetime] = Field(
        default=None,
        description="Absolute time to fire the reminder (UTC)",
    )
    offset_minutes: Optional[int] = Field(
        default=None,
        ge=-10080,  # Max 1 week before
        le=0,
        description="Minutes before due_date to fire (negative value, e.g., -30)",
        examples=[-30, -60, -1440],
    )

    @model_validator(mode="after")
    def validate_fire_time(self) -> "ReminderCreate":
        """Ensure either fire_at or offset_minutes is provided, but not both."""
        if self.fire_at is None and self.offset_minutes is None:
            raise ValueError("Either fire_at or offset_minutes must be provided")
        if self.fire_at is not None and self.offset_minutes is not None:
            raise ValueError("fire_at and offset_minutes are mutually exclusive")
        return self


class ReminderResponse(BaseModel):
    """
    Response schema for reminder data.

    Task Reference: T032
    """

    id: str = Field(
        ...,
        description="Unique reminder identifier",
        examples=["rem_abc123"],
    )
    todo_id: str = Field(
        ...,
        description="Associated todo ID",
    )
    occurrence_id: Optional[str] = Field(
        default=None,
        description="Associated occurrence ID (for recurring todos)",
    )
    fire_at: datetime = Field(
        ...,
        description="When the reminder will fire (UTC)",
    )
    offset_minutes: Optional[int] = Field(
        default=None,
        description="Minutes offset from due date (if set)",
    )
    status: Literal["pending", "sent", "snoozed", "cancelled"] = Field(
        ...,
        description="Current reminder status",
        examples=["pending"],
    )
    sent_at: Optional[datetime] = Field(
        default=None,
        description="When the reminder was sent (if sent)",
    )
    snoozed_until: Optional[datetime] = Field(
        default=None,
        description="Snooze end time (if snoozed)",
    )
    created_at: datetime = Field(
        ...,
        description="When the reminder was created",
    )

    model_config = {"from_attributes": True}


class ReminderSnooze(BaseModel):
    """
    Request schema for snoozing a reminder.

    Task Reference: T032
    """

    snooze_minutes: int = Field(
        ...,
        ge=5,
        le=10080,  # Max 1 week
        description="Minutes to snooze the reminder",
        examples=[15, 30, 60, 1440],
    )


class ReminderListResponse(BaseModel):
    """
    Response schema for listing reminders.

    Task Reference: T032
    """

    items: list[ReminderResponse] = Field(
        ...,
        description="List of reminders",
    )
    total: int = Field(
        ...,
        description="Total number of reminders",
    )
